From 4a8a543a95092eabff3b905a21da73ccf4319b8d Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?S=C3=B6ren=20Tempel?= <soeren+git@soeren-tempel.net>
Date: Sun, 1 May 2022 08:13:37 +0200
Subject: [PATCH] Hotfix for segmentation fault on ARMv6

Without this change, GCC optimizes out the `h & 0x7fffffff` operation
causing symhash to return negative values which are then used to index
vectors which causes a segmentation fault.

The proper fix is: https://github.com/cisco/ChezScheme/issues/622

See https://github.com/cisco/ChezScheme/issues/622#issuecomment-1114195185
---
 c/foreign.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/c/foreign.c b/c/foreign.c
index 1a4dbf7..f696380 100644
--- a/c/foreign.c
+++ b/c/foreign.c
@@ -89,7 +89,7 @@ static iptr symhash(const char *s) {
   iptr n, h;
 
   h = n = strlen(s);
-  while (n--) h = h * multiplier + *s++;
+  while (n--) h = h * multiplier + (signed char)*s++;
   return (h & 0x7fffffff) % buckets;
 }
 
